### PR TITLE
Intro transfer learning for Unit 2

### PR DESCRIPTION
**Pull request by bellabf**
Hi everyone, 

This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: https://github.com/johko/computer-vision-course/issues/64.

It was a missing/abandoned content from the first sprint (the priority was notebooks).

There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch. 

Thanks!

### CODE REVIEW

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+## Transfer learning and Self-training
+
+Transfer learning shines especially when there is not enough labeled data to retrain a model from scratch.  Using our example, we can think that given enough time, a player who attends just a few lessons can learn on their own by playing the instrument without the constant supervision of their professor. Learning, partially or entirely, on your own in deep learning is called self-training. It allows us to train the model using both labeled (the lessons) and unlabeled (the players on their own) to learn the task.  
+
+Although we will not discuss the concept of self-training in this section, we mention it here as a resource to you because when transfer learning does not work and labeled data is scarce, [self-training can be incredibly helpful](https://doi.org/10.48550/arXiv.2006.06882). These concepts are also not mutually exclusive, a seasoned player might need just a couple of lessons to become autonomous in a new instrument training without supervision and, as it turns out, so do our deep learning models.
+
+## Ressources
```
Comment (miniMaddy) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:33]: It should be Resources

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# Transfer Learning
+Before we can dig into the details of what transfer learning and fine-tuning mean for neural networks, let’s take musical instruments as an example. The theremin is an electronic musical instrument that makes an eerie sound, commonly associated with thrillers and horror movies. It is very hard to play because it requires you to move both your hands in the air between two antennae to control the pitch and volume. So hard, that someone invented an instrument called the Tannerin (it is also known as slide Theremin or Eletric-Theremin) that makes a similar sound but it is easier to play.  The player moves the slides at the side of the box to the desired frequency to create a pitch. There is still a learning curve to play it... Well, except if you play the trombone. When you play the trombone, you already know how to use the tannerin slide because it is the same as the telescope slide mechanisms on the trombone. Below you see from left to right: the theremin, tannerin, and the trombon.
```
Comment (miniMaddy) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:2]:
```suggestion
Below you see from left to right: the theremin, tannerin, and the trombon.
```
Please add a comma after Below

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Transfer learning requires that the previous knowledge is “useful” for the other task. Thus, the features we are trying to explore need to be general enough for the new application. If we go back to our musical instrument example, playing the saxophone instead of the trombone is not as helpful in learning how to play the tannerin.  The main skill that gives the trombone player its head start is the intuitive understanding of where the slide should be. 
+
+Yet, the saxophone player is not starting from zero. He is familiar with things like music theory, rhythm, and timing. These general skills give them an edge over someone who never played any instrument at all.  The act of playing an instrument gives all players a general set of skills that are useful across instruments. This generalization across domains (in our example musical instruments) is what makes the model learn much faster as opposed to training from zero.
```
Comment (miniMaddy) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:11]:
```suggestion
Yet, the saxophone player is not starting from zero. He is familiar with things like music theory, rhythm, and timing. These general skills give them an edge over someone who never played any instrument at all.  The act of playing an instrument gives all players a general set of skills that are useful across instruments. This generalization across domains (in our example, musical instruments) is what makes the model learn much faster as opposed to training from zero.
```
Adding a comma

**Comment by miniMaddy** [comment]
Great content @bellabf , I added a few changes. I would also suggest to maybe add the hugging face models (https://huggingface.co/models) link or some other link for pretrained models.

**Comment by bellabf** [comment]
> Great content @bellabf , I added a few changes. I would also suggest to maybe add the hugging face models (https://huggingface.co/models) link or some other link for pretrained models.

Thanks for the quick review and kind words :hugs:! I've addressed all your comments and committed the suggest changes.

**Comment by ATaylorAerospace** [comment]
> Hi everyone,
> 
> This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: #64.
> 
> It was a missing/abandoned content from the first sprint (the priority was notebooks).
> 
> There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch.
> 
> Thanks!

Hello @bellabf ! Do you want us to review/approve now or wait until the realted notebooks are completed?

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Transfer learning requires that the previous knowledge is “useful” for the other task. Thus, the features we are trying to explore need to be general enough for the new application. If we go back to our musical instrument example, playing the saxophone instead of the trombone is not as helpful in learning how to play the tannerin.  The main skill that gives the trombone player its head start is the intuitive understanding of where the slide should be. 
+
+Yet, the saxophone player is not starting from zero. He is familiar with things like music theory, rhythm, and timing. These general skills give them an edge over someone who never played any instrument at all.  The act of playing an instrument gives all players a general set of skills that are useful across instruments. This generalization across domains (in our example, musical instruments) is what makes the model learn much faster as opposed to training from zero.
+
+## Transfer Learning and Fine Tuning
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:13]:
```suggestion
## Transfer Learning and Fine-tuning
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+Let's make a distinction between the concepts we are talking about. The trombone player needs no training to play the tannerin, he already knows how to do it unbeknownst to him. The saxophone player needs some training to fine-tune his skills to play the tannerin. In deep learning terms, what the trombone player does is use a model out-of-the-shelf. This is called transfer learning. The training of a model that needs more time to learn, like our saxophone player, is called fine-tuning.
```
Comment (merveenoyan) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:15]:
```suggestion
Let's make a distinction between the concepts we are talking about. The trombone player needs no training to play the tannerin. He already knows how to do it unbeknownst to him. The saxophone player needs some training to fine-tune his skills to play the tannerin. In deep learning terms, the trombone player uses a model off-the-shelf. This is called transfer learning. The training of a model that needs more time to learn, like our saxophone player, is called fine-tuning. 
```

**Review by merveenoyan — COMMENTED** [review]
Markdown part is great! just left wording nits

File: notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb [hunk 6]
```diff
@@ -0,0 +1,1187 @@
+{
```
Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:1]: <p>we could say "We will now initialize the cross entropy loss function and the SGD optimizer."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:1]: <p>Fine-tuning the Model*</p><p>Also would be nice to have few explanations below the header</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:1]: <p>Same here, would be nice to have few explanations</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/198/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

**Comment by merveenoyan** [comment]
@bellabf thanks a lot for picking up this section 🦸‍♀️

**Comment by bellabf** [comment]
> > Hi everyone,
> > This is high level intro to transfer learning for Unit 2 CNN. I followed the content suggest in this discussion: #64.
> > It was a missing/abandoned content from the first sprint (the priority was notebooks).
> > There are still one or two notebooks from this part that are on being worked on right now, but are still being reviewed in our branch.
> > Thanks!
> 
> Hello @bellabf ! Do you want us to review/approve now or wait until the related notebooks are completed?

Sorry, I missed this comment early. I've sent the .mdx early to review it faster. I've tagged you because you were the person who had made a lot of suggestions in the initial outline discussion. We will focus on reviewing the notebook now! Thanks :smile:

**Review by ATaylorAerospace — COMMENTED** [review]
Hello @bellabf I did not see any additional nits outside of items that @merveenoyan allready noted :-)

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+![Theremin, Tannerin, and the Trombone](
+https://huggingface.co/datasets/hf-vision/course-assets/resolve/d0096005da7fe2eb3bfbd3d0047e9ac7bd499cf0/transfer_learning.png)
+
+In this case, the trombone player has effectively used what he learned by playing the trombone to play the tannerin. He transfers what he learned from one instrument to another. We can use this concept in neural networks as well.  What a neural network learns while classifying dogs or cats can be used to recognize other animals. The explanation for why this works is due to the way networks learn features in the model. That is, the learned feature used to classify a dog, also classify a horse.  We exploit what the model already knows to do different tasks.
+
+Transfer learning requires that the previous knowledge is “useful” for the other task. Thus, the features we are trying to explore need to be general enough for the new application. If we go back to our musical instrument example, playing the saxophone instead of the trombone is not as helpful in learning how to play the tannerin.  The main skill that gives the trombone player its head start is the intuitive understanding of where the slide should be.
```
Comment (johko) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:9]:
```suggestion
Transfer learning requires that the previous knowledge is “useful” for the other task. Thus, the features we are trying to explore need to be general enough for the new application. If we go back to our musical instrument example, playing the saxophone instead of the trombone is not as helpful in learning how to play the tannerin. The main skill that gives the trombone player its head start is the intuitive understanding of where the slide should be. 
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+If the results after fine-tuning the final task are not satisfactory, we still do not need to retrain the entire feature extraction part.  A good compromise is to retrain only the weights of the top layers. In convolutional networks, the higher up a layer is, the more characteristic its features are to the task and dataset. In other words, the features in the first convolutional layers are more generic, while the last layers are more specific. With our player example, this is the equivalent of not wasting time trying to explain music theory to a seasoned saxophonist, but instead just teaching him how to change pitch in the tannerin.
+
+## Considerations on Transfer Learning
+
+Our example also gives us an interesting nuance. The theremin was too hard to play, so they invented an easier instrument that produced the same sound.  The output is nearly the same but needs a lot less training time. For computer vision, we might first do object detection to see where a dog is within an image, and then build a classifier to tell us which breed of dog instead of trying to build a classifier right away.
```
Comment (johko) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:23]:
```suggestion
Our example also gives us an interesting nuance. The theremin was too hard to play, so they invented an easier instrument that produced the same sound. The output is nearly the same but needs a lot less training time. For computer vision, we might first do object detection to see where a dog is within an image, and then build a classifier to tell us which breed of dog instead of trying to build a classifier right away.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+Finally, transfer learning is not a universal performance enhancer. In our example, playing an instrument might help us learn another one, but it might also hinder progress.  There are patterns and vices from one instrument that might slow down the progress of another one. If these vices are deeply entrenched within the player, a novice player might surpass the new player with the same amount of training. If your players are stuck to their vices, it might be time to hire new ones.
```
Comment (johko) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:25]:
```suggestion
Finally, transfer learning is not a universal performance enhancer. In our example, playing an instrument might help us learn another one, but it might also hinder progress. There are patterns and vices from one instrument that might slow down the progress of another one. If these vices are deeply entrenched within the player, a novice player might surpass the new player with the same amount of training. If your players are stuck to their vices, it might be time to hire new ones.
```

File: chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+## Transfer learning and Self-training
+
+Transfer learning shines especially when there is not enough labeled data to retrain a model from scratch.  Using our example, we can think that given enough time, a player who attends just a few lessons can learn on their own by playing the instrument without the constant supervision of their professor. Learning, partially or entirely, on your own in deep learning is called self-training. It allows us to train the model using both labeled (the lessons) and unlabeled (the players on their own) to learn the task.
```
Comment (johko) [inline chapters/en/Unit 2 - Convolutional Neural Networks/intro-transfer-learning.mdx:29]:
```suggestion
Transfer learning shines especially when there is not enough labeled data to retrain a model from scratch. Using our example, we can think that given enough time, a player who attends just a few lessons can learn on their own by playing the instrument without the constant supervision of their professor. Learning, partially or entirely, on your own in deep learning is called self-training. It allows us to train the model using both labeled (the lessons) and unlabeled (the players on their own) data to learn the task.  
```

**Review by johko — COMMENTED** [review]
Nice work on the mdx part. Really well explained with the instrument parallels

**Comment by ratan** [comment]
Nice work @bellabf.
All the minor nits are already mentioned by @merveenoyan @johko. Looks good

**Review by ratan — COMMENTED** [review]
Looks good

**Review by michaelshekasta — APPROVED** [review]
I think that the explanation is very clear
Pay attention - I don't check the notebooks

**Comment by merveenoyan** [comment]
@bellabf can you run the styling scripts and we can merge?

**Comment by aman06012003** [comment]
I have updated the notebook as per the sugesstions from @merveenoyan

**Review by ratan — APPROVED** [review]

**Comment by bellabf** [comment]
> @bellabf can you run the styling scripts and we can merge?

Hi, Merve! I am not sure what/where is the styling scripts. Can you point them out to me? Thank you so much.

**Comment by merveenoyan** [comment]
@bellabf apparently style checks are passing, you just need to add this to toctree and we'll merge 😊

**Comment by merveenoyan** [comment]
@bellabf if you can implement the following changes, the build one will go green 
https://github.com/johko/computer-vision-course/pull/223/commits/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/pull/223/commits/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/pull/223/commits/bd5673b36b8b75a39e941a211ba10c023c349399
https://github.com/johko/computer-vision-course/pull/223/commits/63cabaa80138bc3128b00163980576d58a19ce8f

**Review by merveenoyan — APPROVED** [review]

**Comment by bellabf** [comment]
@merveenoyan Hi, Merve. I was away for these last few days, thank you for taking care of this :hugs:.
